ASoC: SOF: ipc4-topology: Runtime module audio format print rework

ujfalusi opened the pull request:
Hi,

When the audio started we print out here and there the Input and Output or only the Input audio format for the module that is
in path of the stream.
This is mostly OK but provides inconsistent an - in my opinion - hard to grasp information.

The PR reworks how the input/output format is selected and tries to improve the consistency of the prints by applying universal rules:
- Module is not changing the format or it has only input or output _and_ it has single format on in and out side:
```
Audio format for gain.1.1:
```
- Module is not changing the format or it has only input or output _and_ it has multiple formats on either in or out side:
```
Audio format (in/out format index: 2/0) for host-copier.0.playback:
```
- Module is changing format then we print the input and output ones:
- Input if it has single format:
```
Input audio format for host-copier.0.capture:
```
- Input has multiple formats:
```
Input audio format (format index: 0) for host-copier.0.playback:
```
Similar versions for the output format prints.

ujfalusi committed:
ASoC: SOF: ipc4-topology: Use local variables in sof_ipc4_init_input_audio_fmt()

We have local copies of available_fmt->input_pin_fmts and
available_fmt->num_input_formats, use them in the function.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Remove redundant check in sof_ipc4_init_input_audio_fmt()

At label in_fmt the if (pin_fmts_size && i < pin_fmts_size) is guarantied
to be true all the time, drop the check completely.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Simplify match format print in sof_ipc4_init_input_audio_fmt()

Print out the information line for the found input format once to avoid
duplicated prints in case when multiple formats are available.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Use local variables in sof_ipc4_init_output_audio_fmt()

Use local variables for available_fmt->output_pin_fmts and
available_fmt->num_output_formats similarly to the input format selection
to make the two functions easier to understand and help with readability.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Simplify code to deal with process modules without output

Process modules are allowed to have zero outputs, thus zero output formats.

In this case there is no need for complicated if expressions to handle such
cases, we can just use a single if for the number of output formats and
the rest can be simplified.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Concentrate prints inside of sof_ipc4_init_output_audio_fmt()

Similarly to sof_ipc4_init_input_audio_fmt(), move all output format
selection related prints (success or failure) inside of the
sof_ipc4_init_output_audio_fmt() function.

To do this, we need to pass swidget also, like with the input counterpart.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Drop the 'index' from 'Pin index' of format print

Printing the word `index` does not give extra information over printing:
`Pin #%d` which tells that the format line is for a specific pin.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Add helper function to print the module's in/out audio format

Introduce a helper function to print out the audio format(s) used by a
module in a consistent way.

The printed text depends on the module format configuration, taking into
account if they have both input and output support, the format is changed
by the module and the number of formats supported on input/output.

For example, if a module does not change format, there is no point of
printing both in and out format, it is adequate to just state the format
the module is using.

While the function to generate the print is fairly complex (but not too
much), it will create a cleaner experience on the reader side by handling
the filtering of the information and present it in a way that it - I hope -
makes the developer's live a bit more easier when tracking format changes.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -195,7 +195,7 @@ static void sof_ipc4_dbg_audio_format(struct device *dev, struct sof_ipc4_pin_fo
 		struct sof_ipc4_audio_format *fmt = &pin_fmt[i].audio_fmt;
 		dev_dbg(dev,
-			"Pin index #%d: %uHz, %ubit, %luch (ch_map %#x ch_cfg %u interleaving_style %u fmt_cfg %#x) buffer size %d\n",
+			"Pin #%d: %uHz, %ubit, %luch (ch_map %#x ch_cfg %u interleaving_style %u fmt_cfg %#x) buffer size %d\n",
```
commit message: missing word after "as the index is not" ... not what?

ujfalusi replied:
it is a leftover from several attempts to reason and drop thing from the long line.

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2343,47 +2343,55 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
-	out_ref_rate = in_fmt->sampling_frequency;
-	out_ref_channels = SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(in_fmt->fmt_cfg);
-	out_ref_valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(in_fmt->fmt_cfg);
+	/* Configure output audio format only if the module supports output */
```
revisit the commit title

ASoC: SOF: ipc4-topology: Only deal with output format if there is any for process module

??

ASoC: SOF: ipc4-topology: test presence of output format before updating hw_params

or something along those lines.

ujfalusi replied:
it is more than just updating the params as we are also skipping of selecting an output format.
in essence this patch simplifies the complicated if() statements.

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2351,22 +2351,37 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
+						    available_fmt);
+		if (ret < 0)
+			return ret;
+	}
```
how was this tested since we don't have such modules, and the entire flow is probably broken anyways since there's no trigger to propagate.

I am not sure this has any merits for now.

ujfalusi replied:
OK, I will drop this patch, I had it for symmetry, but right, we will fail if num_input_formats == 0

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -203,6 +203,93 @@ static void sof_ipc4_dbg_audio_format(struct device *dev, struct sof_ipc4_pin_fo
 	}
 }
 
+static void
```
commit message: what about the case of a module with an input and no output, e.g. to analyze data?

Why special case the no output configuration?

ujfalusi replied:
We are not special casing things in code, I just mention that process modules can have no outputs.
As per your comment modules w/o input makes no sense as they cannot be triggered, right?

I will re-word the commit message.

plbossart replied:
we had exactly that problem with the tone generator, it was impossible to start...

But that's a framework issue, sooner or later that sort of generator will be needed.

ujfalusi replied:
I had a tone generator in mind when added the patch to handle process module without input, which is dropped now.
My thinking was to prepare the code so if ever this sort of thing is coming along we don't have the kernel-firmware-topology needing update in sync.

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -203,6 +203,93 @@ static void sof_ipc4_dbg_audio_format(struct device *dev, struct sof_ipc4_pin_fo
+			swidget->widget->name);
+	else
+		dev_dbg(dev, "Input audio format (format index: %d) for %s:\n",
+			in_fmt_index, swidget->widget->name);
```
what if num_input_formats == 0?

ujfalusi replied:
See line 253 and 218-235. It will not happen, but we will treat it as the module have identical format and print accordingly.

plbossart reviewed (commented):
Not following a number of comments, nor how this was tested. This looks like speculative preparation for hypothetical usages we can't test. That adds the risk of adding as many issues as problems solved...

ujfalusi commented:
Changes since v1:
- drop patch to support (iow, handle) process modules with only outputs
- Rework the common format printing function to be easier to read and firm rules to handle all possible permutations
- Commit messages updated when requested.

plbossart commented:
SOFCI TEST

plbossart commented:
@bardliao can you take a look?

bardliao commented:
> @bardliao, @kv2019i, @ranj063, any objection to merge this PR?

No, I was approved. :)

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2351,47 +2351,55 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
-	out_ref_channels = SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(in_fmt->fmt_cfg);
-	out_ref_valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(in_fmt->fmt_cfg);
+	/* Configure output audio format only if the module supports output */
+	if (available_fmt->num_output_formats) {
```
@ujfalusi just wondering if num_output_pins is a better indicator of whether a module supports output or not.

ujfalusi replied:
I'm trying to avoid functional changes and the original code also used `num_output_formats`, I would not touch this for now, but revisit later?

ranj063 replied:
alright then. this is good to go
